content/docs/rfcs/66/README.md

```diff
@@ -0,0 +1,32 @@
+---
+slug: 66
+title: 66/WAKU2-METADATA
+name: Waku Metadata Protocol
+status: raw
+editor: Alvaro Revuelta <alrevuelta@status.im>
+contributors:
+---
+
+# Metadata Protocol
+
+Waku specifies a req/resp protocol that provides information about the node's medatadata.
+
+
+## Protocol id
+
+`/vac/waku/metadata/1.0.0`
+
+## Request
+
+```proto
+message WakuMetadataRequest {
```

Comment:
Any sense in including the desired metadata set here (similar to how https://github.com/libp2p/specs/blob/master/connections/README.md#multistream-select works). In other words, as a requester I include the capabilities that I want in the request and the responder responds with the subset of my requested capabilities that it supports. Advantage here is that it's more of a negotiation and will make this protocol more adaptable to be used for different capabilities (such as store, etc.). For now this would only be used for network ID "negotiation", but allows both the requester and responder to know if they are interested in this connection.

Reply:
Maybe I misunderstood multistream-select, but doesn't it work in a way that you ask if the peer supports particular protocol and get back basically true (protocol id) or false (na) ?

So is the idea for requester to basically send "expected" response and then responder kind of "confirms" the requests and maybe provides some details?

Reply:
Use cases of this protocol beyond networkId (and perhaps shards) are a bit unclear to me. While I agree with anticipating to the future, I'm not sure we have to solve all capability related things right now. Mainly because we haven't even defined yet what capabilities we have (besides the protocols themselves: store, lp, filter, px, relay). Shards are the only capability that are clear imho.

Regarding multistream-select, not sure I understand your suggestion. Haven't worked with it. Are you suggesting to use that instead of adding a new metadata protocol?

> as a requester I include the capabilities that I want in the request and the responder responds with the subset

since we dont have many capabilities, responding with all your capabilities is simpler or? I mean, if we had thousands of capabilities it would be better to reply with just the common ones, but does it make sense now?

Reply:
Indeed, we shouldn't overdesign for capabilities that we don't yet have or aren't sure how it will work.
However, I think there may be value in viewing this as a negotiation protocol similar to multistream-select even in its simplest form (it would still be a new protocol, multistream-select is just an example). In other words, the requester sends the list of capabilities/metadata for which a connection would be interesting to itself. The responder responds with the (sub) set that it supports. For network ID this would simply work:
- requester sends "desired" network ID, shards
- responder responds with same network ID and the shards it does support.

Two immediate advantages:
- this would allow the responder to disconnect after this exchange without having to initiate a similar req-resp to "know" if the connection is interesting from its perspective.
- both the requester and responder only divulge the information they consider necessary for a successful connection. No advertising of unrelated capabilities.

However, this becomes much more interesting when metadata becomes more complex. Even if we have not designed this yet, an example: I'm interested in store protocol for shard 7 for the last 24 hours. A store node supports relay on shard 9, have stored messages for shard 7 for 12 hours and for shard 9 for 3 days. This is of course artificial, but shows how quickly metadata advertising can become complex, especially when multiple shards are involved. If we switch it to the requester simply requesting "do you have store for network 1, shard 7, 24 hours" the responder can simply respond "network 1, shard 7, 12 hours".

The advantages:
- much more flexibility: it's complex to advertise "all" metadata not knowing what parts is interesting to clients, but (usually) simple to formulate and respond to requests (i.e. negotiation)
- in certain cases, the responding node may choose to start supporting ad-hoc capabilities based on what requesters are interested in. If e.g. a node is interested in filter services on shard 7, I could choose to subscribe to shard 7 at the point I receive the request and respond with a "yes" for shard 7. This will become especially useful if we start incentivising service protocols.

Happy to do something simple for now, though if we believe this _direction_ might be valuable in future, I think it's better to have it function as a negotiation protocol early on already.

Reply:
how about this?
d381ee1bbdaf033a4561eda773048117e2be731d